### PR TITLE
Step 4

### PR DESCRIPTION
**Pull request by tonklable**
## What

Established database
Migrate GET function to db
Migrate GET :itemId function to db
Migrate POST function to db
Create GET search function

## CHECKS :warning:

Please make sure you are aware of the following.

- [ ] **The changes in this PR doesn't have private information

### CODE REVIEW

File: go/app/main.go [hunk 1]
```diff
@@ -59,48 +61,80 @@ func addItem(c echo.Context) error {
...
+	if err != nil {
+		c.Logger().Errorf("Error opening database, %v", err)
+		res := Response{Message: "Error opening database"}
+		return c.JSON(http.StatusInternalServerError, res)
+	}
+
+	defer db.Close()
+
+	// Insert item into database
+	result, err := db.Exec("INSERT INTO items (name, category, image) VALUES (?, ?, ?)", newItem.Name, newItem.Category, newItem.ImageFilename)
```
Comment (karuna) [inline go/app/main.go:76]: question mark is good because it's prepared statement and can avoid sql injection. 👍

File: go/app/main.go [hunk 2]
```diff
@@ -59,48 +61,80 @@ func addItem(c echo.Context) error {
...
-		c.Logger().Errorf("Error saving items to JSON: %s", err)
-		res := Response{Message: "Error saving items to JSON"}
+		c.Logger().Errorf("Error getting last insert ID, %v", err)
+		res := Response{Message: "Error getting last insert ID"}
 		return c.JSON(http.StatusInternalServerError, res)
 	}
 
-	c.Logger().Infof("Receive item: %s", newItem)
-	res := Response{Items: existingItems.Items}
+	newItem.Id = int(lastInsertID)
```
Comment (karuna) [inline go/app/main.go:91]: is this `int()` typecasting safe?

File: go/app/main.go [hunk 3]
```diff
@@ -115,32 +149,82 @@ func getItemDetail(c echo.Context) error {
...
+		res := Response{Message: "Error opening database"}
 		return c.JSON(http.StatusInternalServerError, res)
 	}
-	defer jsonFile.Close()
 
-	jsonData, err := ioutil.ReadAll(jsonFile)
+	defer db.Close()
+
+	// Query database
+	row, err := db.Query("SELECT id, name, category, image FROM items WHERE id LIKE ? OR name LIKE ? OR category LIKE ? OR image LIKE ?", "%"+keyword+"%", "%"+keyword+"%", "%"+keyword+"%", "%"+keyword+"%")
```
Comment (karuna) [inline go/app/main.go]: have you check if this is safe from sql injection?

File: go/go.sum [hunk 4]
```diff
@@ -11,6 +11,10 @@ github.com/mattn/go-colorable v0.1.11 h1:nQ+aFkoE2TMGc0b68U2OKSexC+eq46+XwZzWXHR
 github.com/mattn/go-colorable v0.1.11/go.mod h1:u5H1YNBxpqRaxsYJYSkiCWKzEfiAb1Gb520KVy5xxl4=
 github.com/mattn/go-isatty v0.0.14 h1:yVuAays6BHfxijgZPzw+3Zlu5yQgKGP2/hcQbHb7S9Y=
 github.com/mattn/go-isatty v0.0.14/go.mod h1:7GGIvUiUoEMVVmxf/4nioHXj79iQHKdU27kJ6hsGG94=
+
```
Comment (karuna) [inline go/go.sum]: weird, did you manually edit this file? there should be no new line here

File: go/app/main.go (hunk 2 above)
Comment (tonklable) [reply go/app/main.go:91]: Yes, I want to make sure the id is int in the newItem

**Pull request merged**